proposals/0127-get-sysvar-syscall.md

```diff
@@ -0,0 +1,227 @@
+ * @param offset    The offset to start copying data from.
+ * @param length    The length of data to copy, in bytes.
+ * @return          A 64-bit unsigned integer error code:
+ *                    - 0 if the operation is successful.
+ *                    - Non-zero error code.
+ *
+ * If the operation is not successful, data will not be written to the
+ * provided VM memory address.
+ */
+uint64_t sol_get_sysvar(
+  /* r1 */ void const * sysvar_id,
+  /* r2 */ uint8_t *    var_addr,
+  /* r3 */ uint64_t     offset,
+  /* r4 */ uint64_t     length,
+);
+
+```
+
+### Control Flow
+
+The syscall aborts the virtual machine if any of these conditions are true:
+
+- Not all bytes in VM memory range `[sysvar_id, sysvar_id + 32)` are readable.
+- Not all bytes in VM memory range `[var_addr, var_addr + length)` are writable.
+- `offset + length` is not in `[0, 2^64)`.
+- `var_addr + length` is not in `[0, 2^64)`.
+- Compute budget is exceeded.
+
+The syscall returns `1` if the sysvar is not present in the Sysvar Cache.
+
+The syscall returns `2` if `offset + length` is greater than the size of the
+sysvar.
+
+Otherwise, the syscall writes the sysvar data to the virtual memory at
+`var_addr` and returns `0`.
+
+### Compute Unit Usage
+
+The syscall will always consume the same amount of CUs regardless of
+control flow.
+
+CU usage is proportional to the `length` parameter.
+
+```
+sysvar_base + (32 / cpi_per_u) + max(mem_op_base, (length / cpi_per_u))
+```
+
+- `sysvar_base`: Base cost of accessing a sysvar.
+- `cpi_per_u`: Number of account data bytes per CU charged during CPI.
+
+### Supported Sysvars
+
+The proposed unified syscall interface will support all non-deprecated sysvars
+in the Sysvar Cache.
+
+Unless specified otherwise, any new sysvars added to the Sysvar Cache in the
+future also become accessible through this syscall.
+
+Existing syscalls to fetch sysvar data will be deprecated in favor of the new
+unified syscall proposed herein.
+
+The supported list of sysvars for the proposed syscall will be as follows:
+
+- `SysvarC1ock11111111111111111111111111111111`
+- `SysvarEpochRewards1111111111111111111111111`
+- `SysvarEpochSchedu1e111111111111111111111111`
+- `SysvarLastRestartS1ot1111111111111111111111`
+- `SysvarRent111111111111111111111111111111111`
+- `SysvarS1otHashes111111111111111111111111111`
+- `SysvarS1otHistory11111111111111111111111111`
+- `SysvarStakeHistory1111111111111111111111111`
+
+Sysvar APIs at the SDK level are responsible for defining exactly how the data
+can be accessed using the new syscall. For example, multiple SDKs may choose to
+allow users to access one element from a list-based sysvar at a time, while only
+a few may offer lookups or binary searches on the data.
+
+## Impact
+
+As mentioned as partial motivation for this proposal, the new syscall would
+make changing and adding new sysvars in the future much easier, which would be
+positively impactful to contributors.
+
```

Comment:
Another minor and not so obvious impact is that this changes the verification of supported sysvar related syscalls from verification time to runtime.

Reply:
Added a note on this in the section you've commented.